Make Compiler and CodeGen objects smaller

mikedn opened the pull request:
This brings down the size of `Compiler` from around 7.2 kbytes to around 1.9 kbytes. Some data members have moved to `CodeGen` but then the `siLatestTrackedScopes` change also brings down the size of `CodeGen` from around 5 kbytes to around 1 kbyte.

This decreases the JIT memory usage by 11.5% (14.9% with minopts) for crossgen corelib.

Mem stats diff: https://gist.github.com/mikedn/eabe00eb09a785a4d2fe24d07f80d820

This also results in 0.5% decrease in instructions retired: https://1drv.ms/x/s!Av4baJYSo5pjgr8Qa-DKqXP_4mJE6g Most of this comes from `siLatestTrackedScopes`, we no longer zero out a 512 element array. Though it's seems a bit too good to be true, maybe PIN gets confused by the `memset` call or something.

CarolEidt commented on src/jit/lclvars.cpp:
```diff
@@ -3507,9 +3509,14 @@ void Compiler::lvaSortByRefCount()
 
+    if (lvaTrackedToVarNum == nullptr)
+    {
+        lvaTrackedToVarNum = static_cast<unsigned*>(compGetMemArray(lclMAX_TRACKED, sizeof(unsigned), CMK_LvaTable));
```
Nice. This means we won't even use this space for minopts.

mikedn replied:
I hope so. I haven't actually verified that minopts doesn't attempt to sort variables :smile:

There may be more room for improvement around sorting. Maybe we don't need both `lvaTrackedToVarNum` and `lvaRefSorted`, they're pretty much the same thing.

BruceForstall replied:
Seems like we should only need to allocate enough space for the tracked variables, not `lclMAX_TRACKED` space. Currently, we compute lvaTrackedCount below, but it seems like we could compute the total number of tentative tracked vars in the loop above, then cap it at `lclMAX_TRACKED`.

Seems like if we're using "ShortLong" bitsets, and dynamic allocation for things like this, we don't even require `lclMAX_TRACKED` to be constant anymore (though I can't think why we would change that).

BruceForstall replied:
It would be nice to have a function:
```
unsigned lvaTrackedIndexToVarNum(unsigned index)
{
  asssert(index < lvaTrackedCount);
  return lvaTrackedToVarNum[index];
}
```
just to get that assert.

mikedn replied:
> Seems like we should only need to allocate enough space for the tracked variables, not lclMAX_TRACKED space

This is something I hope to look at in the future. It's probably a bit more complicated because we sort more than once and new variables can be added. We'll probably need to over allocate initially to avoid having to discard the existing array and allocate a new one.

There's also the above mentioned issue that `lvaTrackedToVarNum` and `lvaRefSorted` are redundant, one is enough. And `lvaRefSorted` is allocated every time sorting is done.

And then the use of `qsort` is questionable, its reliance on C-style callbacks makes it kind of expensive, especially on Windows with VC++'s `/cf:guard`.

> It would be nice to have a function:

Yes. Actually more than that:
* `LclVarDsc* lvaGetLclVar(unsigned lclNum);`
* `LclVarDsc* lvaGetLclVarByTrackedIndex(unsigned index);`
* and maybe `unsigned lvaGetLclNum(LclVarDsc* varDsc);`

I find it surprising that in 2018 one can still find code like 
```C++
        unsigned varNum = lvaTrackedToVarNum[deadVarIndex];
        varDsc          = lvaTable + varNum;
```
in a reasonably important codebase.

mikedn replied:
> Nice. This means we won't even use this space for minopts.

@CarolEidt This doesn't seem to be the case. It looks like in minopts we sort twice - once from `lvaMarkLocalVars` and once when doing liveness before LSRA. Without minopts a 3rd sorting happens before SSA (for liveness).

CarolEidt replied:
Interesting. I thought that we didn't track lclVars in minopts, but it's just that we generally don't enregister them because `opts.compFlags & CLFLG_REGVAR` is false. I wonder why this is ... perhaps this is one of those things that would be beneficial to turn off for Tier 0.

CarolEidt commented on src/jit/compiler.h:
```diff
@@ -7288,56 +7199,66 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+    SIMDHandlesCache* m_simdHandleCache;
+
     // Get the handle for a SIMD type.
     CORINFO_CLASS_HANDLE gtGetStructHandleForSIMD(var_types simdType, var_types simdBaseType)
```
I made this comment earlier, but for some reason github thinks it's outdated:
I think this method should check whether `m_simdHandles` is null, and return `NO_CLASS_HANDLE` in that case.

mikedn replied:
Hmm, I've looked at this when I added `m_simdHandles`. It seems to me that it is impossible to call this function without ever calling `getBaseTypeAndSizeOfSIMDType`. But then maybe it would be possible in the future to encounter SIMD types without actually importing them?

CarolEidt replied:
I agree that it doesn't seem possible *now* but callers should be resilient to getting `NO_CLASS_HANDLE`  back, and it's basically just defensive programming.

mikedn replied:
OK, I'll change it. I was thinking that if this ever happens in the future then maybe we should create the cache here. But that wouldn't work anyway since we don't have any type name to lookup. So returning `NO_CLASS_HANDLE` does make sense if the compiler suddenly generates SIMD types internally.

In fact it would probably be great if SIMD types behaved more like primitive types, without this handle wizardry beyond import.

CarolEidt commented:
@AndyAyersMS @briansull it would be good to get some "front-end" eyes on this as well I think.

AndyAyersMS commented on src/jit/importer.cpp:
```diff
@@ -17385,26 +17385,39 @@ void Compiler::impImport(BasicBlock* method)
+        if (inlineRoot->impStkSize < impStkSize)
+        {
+            inlineRoot->impStkSize              = impStkSize;
+            inlineRoot->verCurrentState.esStack = new (this, CMK_ImpStack) StackEntry[impStkSize];
```
Wouldn't we need this to use the `inlineRoot`'s allocator?

mikedn replied:
Hmm, good question. Both compilers use the same arena allocator so as it is correct. But it probably skews mem stats as each compiler has its own. That's actually bogus anyway.

AndyAyersMS replied:
Ah never mind, looks like we already do...

mikedn commented:
> I'm not sure it totally makes sense to put the spill temps on RegSet

Yep, but it's certainly better than having them in `Compiler`. `RegSet` is a significant user of temps and it's already in codegen so putting temps there resulted in the fewest changes (especially that most tmp functions were already in regset.cpp). I plan to take a closer look at temps after this, it looks rather convoluted. For example, do we still need the `tmpListBeg`/`tmpListNxt` now that temps are preallocated?

In general, this change is just a step along the way (and hopefully in the right direction). Ideally we'd split the giant `Compiler` class into smaller class - `Importer`, various optimization classes (`CopyProp`, `Cse` etc.) so we don't have this fixed multi-kilobyte cost when doing inlining or minopts. But it's a ton of work (especially on the review side, moving things around isn't that hard after all) so I picked up a few changes that are relatively simple.

The SIMD handle cache is a good example. It probably makes sense to try to move some of the code that uses it from `Compiler` to `SIMDHandleCache`. That can be done at a later time. Perhaps we could also figure out if this cache can perhaps be global rather than being associated with a single method compilation.

mikedn commented on src/jit/compiler.hpp:
```diff
@@ -3066,7 +3066,7 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 /*****************************************************************************/
 
-/* static */ inline unsigned Compiler::tmpSlot(unsigned size)
+/* static */ inline unsigned RegSet::tmpSlot(unsigned size)
```
Yep, I plan to take a look at temps after this. There may be other things to clean up.

CarolEidt commented:
> do we still need the tmpListBeg/tmpListNxt now that temps are preallocated?

Probably not; I agree that the management of temps needs to be reviewed/redesigned now that we no longer have to worry about `LEGACY_BACKEND`.

Totally agree that this is a great step along the way - thanks!

mikedn commented:
@AndyAyersMS Seems like the SIMD handle cache could be shared by inlinee compilers.

It is shared now, see `getBaseTypeAndSizeOfSIMDType`. I'm actually wondering if it shouldn't be shared between multiple root compilers, I presume that these handles do not change.

BruceForstall commented on src/jit/lclvars.cpp:
```diff
@@ -3507,9 +3509,14 @@ void Compiler::lvaSortByRefCount()
 #ifdef DEBUG
     // Re-Initialize to -1 for safety in debug build.
-    memset(lvaTrackedToVarNum, -1, sizeof(lvaTrackedToVarNum));
+    memset(lvaTrackedToVarNum, -1, lclMAX_TRACKED);
```
this needs to be:
```
memset(lvaTrackedToVarNum, -1, lclMAX_TRACKED * sizeof(unsigned));
```

mikedn replied:
Whoops, will fix. Thanks!

mikedn commented:
OS X build failed with a few "out of disk space" errors: https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_checked_osx10.12_innerloop_prtest/4404/

@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

mikedn commented:
Fixed the `memset` issue.

I've done a quick test to see what happens if we allocate a properly sized `lvaTrackedToVarNum` - we gain another 2.5% reduction in memory usage. Avoiding allocating a new array every time variables are sorted should provide even more. Definitely something to look at.

One of the things I like most about these changes is that the needle finally moves for methods that need less than 32k (a common size for L1 cache) to compile. We started at 0%
```
     17 ..      32 ===>      10 count (  0% of total)
```
This PR as is brings this to 7%:
```
     17 ..      32 ===>    2201 count (  7% of total)
```
and experimental `lvaTrackedToVarNum` improvements add another 7%
```
     17 ..      32 ===>    4194 count ( 14% of total)
```

briansull commented on src/jit/optimizer.cpp:
```diff
@@ -1148,20 +1141,29 @@ bool Compiler::optRecordLoop(BasicBlock*   head,
+        for (unsigned j = optLoopCount; j > loopInd; j--)
+        {
+            optLoopTable[j] = optLoopTable[j - 1];
+        }
```
Is this an actual code behavior change?
It would be best to keep this checkin to things that don't cause any behavior changes.
And use a seperate change for anything that could cause some code diffs. (Even if it is a bug fix)

mikedn replied:
Hmm, there should not be any behavior change. Are you referring to the fact that this loop no longer executes when the loop table is nullptr? That's because there's no loop in the table.

briansull replied:
It looked like your were adding new code here, but if you are just reordering stuff that is OK.

BruceForstall commented:
@mikedn Sorry for the slow response. Can you fix the merge conflict?

mikedn commented:
> Sorry for the slow response. Can you fix the merge conflict?

@BruceForstall No problem. I'll wait for the CompAllocator PR to be merged first as this PR introduces new allocation sites (using `compGetMemArray`, not sure what I was thinking...) and it will likely conflict again.

mikedn commented:
I've rebased and replaced `compGetMemArray` uses. 

Let me know if you want this squashed. The last 3 commits should definitely be squashed but the rest are rather unrelated so maybe they should stay as is, up to you.

AndyAyersMS commented:
Caching handles across jit requests is potentially tricky since classes can be unloaded. SIMD handles, I suppose, are unlikely to ever get unloaded, though. Also, there are callbacks from the VM to jit to deal with jit-side cache invalidation, -- `isCacheCleanupRequired` and `clearCache` -- but I don't know if they're implemented properly on either side.

AndyAyersMS commented:
Another thought, perhaps for a follow-on change -- seems like we should consider reordering the compiler members perhaps, sorting by frequency or something (or shuttling larger or related things off to child objects). Could improve cache locality a bit, and reduce code size too.

Maybe a good first step here is finding a way to properly measure the frequency of access.

mikedn commented:
> Caching handles across jit requests is potentially tricky since classes can be unloaded.

Yep, that's why this is currently limited to inlining. Doing more not only that requires some investigation on the VM side but also drags multithreading into the mix and can possible create some reliability issue. JIT's handling of errors such as `noway_assert` relies on the fact that it can throw away everything and start from scratch. If we now add caches into the mix we may have to deal with corrupted caches.

> seems like we should consider reordering the compiler members perhaps, sorting by frequency or something (or shuttling larger or related things off to child objects). Could improve cache locality a bit, and reduce code size too.

Maybe, but they're probably more or less grouped today (e.g. lva members are more or less grouped together) so it's unlikely that changing the order would provide significant improvements in terms of cache usage. 

IMO the biggest problem with large objects like `Compiler` is that they mess up memory statistics. You look at the histogram and wonder how come only 10 methods need less than 32 kbyte memory but then you discover that a lot of that allocated memory is not actually used so it has no real impact on cache usage. Splitting up the `Compiler` class would help here, you'd end up with smaller objects and many of them could be allocated on the stack (e.g. like `RangeCheck` is today). Too bad that this requires a ton of work...

mikedn commented:
In case someone is curious how the `Compiler` class looks after this change: https://gist.github.com/mikedn/7bc1d5442845fab5a1b330d1fc3dd793

AndyAyersMS commented:
Interesting -- there are at least 100 bytes of padding voids in there, probably quite a bit more (got tired of counting them all). We seem to have a predilection for putting pointers just after bools:
```
04E0	|	bool eeInfoInitialized;
04E8	|	Compiler::VirtualStubParamInfo* virtualStubParamInfo;
```
I would guess most of the bools could be profitably bitfielded and adjacently packed and that would get rid of most of the padding and compress the bools down nicely too.

So maybe with a bit of tinkering we could shrink the compiler instance from 1976 bytes to maybe ~1800?

BruceForstall commented:
Since there is only one Compiler object (except for inlinee compiles), we haven't worked ever on reducing padding. IMO, it doesn't seem worth it. I would think converting bool fields to bitfields wouldn't be helpful if access is slower. Certainly, within each logical section of Compiler, we should do appropriate sorting-by-size. But more than that seems hard to justify.
